### DISCUSSION TITLE
How to read packets?

### DISCUSSION THREAD

**Discussion by orangepineapple** [ANLAB-KAIST/KENSv3#147] [Project Q&A 2. Transport Layer (TCP)]
I dont think I understand what we are supposed to do with the packets. Is there a pre-defined structure for the packets, for example, how do I know if a packet is an ACK or a Sync, or just some generic data? (since its just a bunch of bytes) 

In this PA we don't have send() or recv() so do we just assume that every packet is only either an ack or a sync?

Also in the docs for packet.read() it states that data is the destination so is the packet just the header? 

What type are we supposed to cast the data field of the packet to? 

 or do we define all of that information ourselves?

**Comment by orangepineapple** [comment]
follow up: Is it correct to assume we are supposed to define our own system for packets for example custom struct with all the required values that we put into a packet

**Comment by Pusnow** [comment]
> I dont think I understand what we are supposed to do with the packets. Is there a pre-defined structure for the packets, for example, how do I know if a packet is an ACK or a Sync, or just some generic data? (since its just a bunch of bytes)

You should use [`readData()`](https://github.com/ANLAB-KAIST/KENSv3/blob/eba97b6c81103e365e14e3c13194ccbc2a5966f3/include/E/Networking/E_Packet.hpp#L85-L91) method to get data from `struct Packet`, and then you can parse the data.
For parsing, you can use your own structs as well as Linux's structs including `struct ip` in `netinet/ip.h` and `struct tcphdr` in `netinet/tcp.h`.

> In this PA we don't have send() or recv() so do we just assume that every packet is only either an ack or a ~sync~ syn?

No, there are also `SYN/ACK`, `FIN`, `FIN/ACK` packets. 

> Also in the docs for packet.read() it states that data is the destination so is the packet just the header?

Meaning of the data destination is that the method returns data to the address pointed to by `*data`. It also stores payload.

> What type are we supposed to cast the data field of the packet to?

Your choice. `unsigned char*` would be a great option.


> or do we define all of that information ourselves?

Yes, you can define struct, class, functions for them as long as all implementations are stored in `TCPAssignment.hpp` and `TCPAssignment.cpp`.

**Comment by Pusnow** [comment]
> follow up: Is it correct to assume we are supposed to define our own system for packets for example custom struct with all the required values that we put into a packet

Yes, you can define your own structs, classes, and functions.
